# Pokemon Database Analysis and Team Generator

This notebook is an example of use of Python's Pandas library.<br>
It's objective is to explore the possibilities of the library while also serving as programming portfolio.<br>
<br>
This application analyzes a Pokemon dataset extracting informationg about it's stats, types and generations so they<br>
can be used in team generation functions.<br>
<br>
The dataset used was obtained from this repository: https://github.com/KeithGalli/pandas<br>
which also features a great pandas guide<br>
<br>
Without further ado, let's catch'em all!

### 1 - Obtaining and Exploring Dataset

First of all, the dataset must be imported. After that i'll give a glance at some rows.

In [394]:
import pandas as pd

pokemon_df = pd.read_csv("pokemon_dataset.csv")

In [317]:
pokemon_df.head()

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

In [318]:
pokemon_df.tail()

#                 Name   Type 1 Type 2  HP  Attack  Defense  Sp. Atk  \
795  719              Diancie     Rock  Fairy  50     100      150      100   
796  719  DiancieMega Diancie     Rock  Fairy  50     160      110      160   
797  720  HoopaHoopa Confined  Psychic  Ghost  80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic   Dark  80     160       60      170   
799  721            Volcanion     Fire  Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True

By looking at the first and last 5 rows of the table and its headers I can try to deduce a few things:<br>
1. There are a lot of pokemon here! Six generations of monsters and I barely know the third one
2. By quickly checking in a trusted source, I know that Mega Pokemon are actually a temporary power, so they don't exist in the wild naturally
3. Legendary Pokemon are also included in the list, which is obvious by the last column "Legendary"
4. There isn't any information about evolutions as shown by the first three Pokemon
5. Beyond Mega Evolutions, there seems to be other alternate powers as well, as shown by the numbers 797 and 798 which share the same ID
<br>

My objective is to generate some random teams of six pokemon, but there are some rules to follow:<br>
1. Legendary Pokemon are really strong, very special. I can't include them in the teams, I'm a fair player
2. Mega Evolutions and Special Pokemons don't occur naturally, so I can't catch them and by consequence can't build a team with them, so they are also out

With these starting two rules i'll filter my dataset accordingly

In [319]:
legendary_df = pokemon_df[(pokemon_df["Legendary"] == True) & (~pokemon_df["Name"].str.contains("Mega "))]

First five rows of the Legendary dataset

In [320]:
legendary_df.head()

#      Name    Type 1  Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  \
156  144  Articuno       Ice  Flying   90      85      100       95      125   
157  145    Zapdos  Electric  Flying   90      90       85      125       90   
158  146   Moltres      Fire  Flying   90     100       90      125       85   
162  150    Mewtwo   Psychic     NaN  106     110       90      154       90   
262  243    Raikou  Electric     NaN   90      85       75      115      100   

     Speed  Generation  Legendary  
156     85           1       True  
157    100           1       True  
158     90           1       True  
162    130           1       True  
262    115           2       True

In [321]:
mega_df = pokemon_df[(pokemon_df["Name"].str.contains("Mega "))]

First five rows of the Megas dataset

In [322]:
mega_df.head()

#                       Name Type 1  Type 2  HP  Attack  Defense  \
3    3      VenusaurMega Venusaur  Grass  Poison  80     100      123   
7    6  CharizardMega Charizard X   Fire  Dragon  78     130      111   
8    6  CharizardMega Charizard Y   Fire  Flying  78     104       78   
12   9    BlastoiseMega Blastoise  Water     NaN  79     103      120   
19  15      BeedrillMega Beedrill    Bug  Poison  65     150       40   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
3       122      120     80           1      False  
7       130       85    100           1      False  
8       159      115    100           1      False  
12      135      115     78           1      False  
19       15       80    145           1      False

Technical note for the record: I know that there is a Pokemon called "Meganium" so i've filtered for the substring "Mega "<br>

In [323]:
common_df = pokemon_df[(pokemon_df["Legendary"] == False) & (~pokemon_df["Name"].str.contains("Mega "))]

First five rows of the Common dataset

In [324]:
common_df.head()

#        Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  1   Bulbasaur  Grass  Poison  45      49       49       65       65     45   
1  2     Ivysaur  Grass  Poison  60      62       63       80       80     60   
2  3    Venusaur  Grass  Poison  80      82       83      100      100     80   
4  4  Charmander   Fire     NaN  39      52       43       60       50     65   
5  5  Charmeleon   Fire     NaN  58      64       58       80       65     80   

   Generation  Legendary  
0           1      False  
1           1      False  
2           1      False  
4           1      False  
5           1      False

Now i'll check for consistency of these datasets just to be sure<br>
For the sake of clarity I now should have three datasets:
1. Legendaries: Should include only legendary non-mega pokemon
2. Megas: Should include all mega pokemon, common or legendary
3. Common: Should include all non-legendary and non-mega pokemon

In [325]:
print("Total Legendaries:\t", len(legendary_df.index))
print("Total Megas:\t\t", len(mega_df.index))
print("Total Common:\t\t", len(common_df.index))
print("")
print("Total Pokemon:\t\t", len(pokemon_df.index))

if len(pokemon_df.index) == len(legendary_df.index) + len(mega_df.index) + len(common_df.index):
    print("Total Quantities Match!")

Total Legendaries:	 59
Total Megas:		 48
Total Common:		 693

Total Pokemon:		 800
Total Quantities Match!


Just to have a greater certainty, i'll cross these datasets in a "inner join". The resulting datasets should be empty.

In [326]:
lc_test_df = pd.merge(legendary_df, common_df, on=['Name'], how='inner')
lm_test_df = pd.merge(legendary_df, mega_df, on=['Name'], how='inner')
cm_test_df = pd.merge(common_df, mega_df, on=['Name'], how='inner')

In [327]:
print("Cross Legendary X Common:\t", len(lc_test_df.index))
print("Cross Legendary X Mega:\t\t", len(lm_test_df.index))
print("Cross Common X Mega:\t\t", len(cm_test_df.index))

Cross Legendary X Common:	 0
Cross Legendary X Mega:		 0
Cross Common X Mega:		 0


Now i'm sure that my subsets are correct!

### 2 - Data Analysis

In this section I'll look more closely to the stats and types of Pokemon.<br>
I will only deal with the Common dataset for now. The Legendary dataset will be analyzed latter and the Mega dataset is not useful for me.

Summary description of the Common dataset<br>

In [328]:
common_df.describe()

#          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  693.000000  693.000000  693.000000  693.000000  693.000000  693.000000   
mean   359.907648   66.594517   72.825397   69.411255   66.174603   67.194805   
std    210.023306   25.184437   27.878345   28.445003   26.710011   25.107707   
min      1.000000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    178.000000   50.000000   52.000000   50.000000   45.000000   50.000000   
50%    356.000000   64.000000   70.000000   65.000000   61.000000   65.000000   
75%    544.000000   78.000000   90.000000   85.000000   85.000000   80.000000   
max    715.000000  255.000000  165.000000  230.000000  150.000000  230.000000   

            Speed  Generation  
count  693.000000  693.000000  
mean    63.932179    3.337662  
std     26.532503    1.686819  
min      5.000000    1.000000  
25%     45.000000    2.000000  
50%     60.000000    3.000000  
75%     84.000000    5.000000  
max    160.000000    6.000000

Before analyzing this description, I'll register some questions I would like answered at first. More may appear latter.
1. What are the max and min values of each stat?
2. How many different generations exist?
3. How many types exist?
4. How many Pokemon by generation exist?

**What are the max and min values of each stat?**<br>
This already have been answered in the description

In [329]:
max_hp = common_df["HP"].max()
min_hp = common_df["HP"].min()

max_attack = common_df["Attack"].max()
min_attack = common_df["Attack"].min()

max_defense = common_df["Defense"].max()
min_defense = common_df["Defense"].min()

max_spatk = common_df["Sp. Atk"].max()
min_spatk = common_df["Sp. Atk"].min()

max_spdef = common_df["Sp. Def"].max()
min_spdef = common_df["Sp. Def"].min()

max_speed = common_df["Speed"].max()
min_speed = common_df["Speed"].min()

values = {"Value": ["max","min"], 
          "HP": [max_hp, min_hp], 
          "Attack": [max_attack, min_attack], 
          "Defense": [max_defense, min_defense], 
          "Sp. Atk": [max_spatk, min_spatk], 
          "Sp. Def": [max_spdef, min_spdef], 
          "Speed": [max_speed, min_speed]}

stat_values_df = pd.DataFrame(values)

In [330]:
stat_values_df

Value   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed
0   max  255     165      230      150      230    160
1   min    1       5        5       10       20      5

**How many different generations exist?**

In [331]:
print("Generations:", len(common_df["Generation"].unique()))

Generations: 6


**How many types exist?**

In [332]:
print("Types:", len(common_df["Type 1"].unique()))
print(*common_df["Type 1"].unique(), sep = ", ")

Types: 18
Grass, Fire, Water, Bug, Normal, Poison, Electric, Ground, Fairy, Fighting, Psychic, Rock, Ghost, Ice, Dragon, Dark, Steel, Flying


**How many Pokemon by generation exist?**

In [333]:
common_by_generation_df = common_df.groupby(by=["Generation"]).count().drop(columns=["Name", "Type 1", "Type 2", "HP", "Attack", "Defense", "Sp. Atk", "Sp. Def", "Speed", "Legendary"]).rename(columns={"#":"Number"})

In [334]:
common_by_generation_df

Number
Generation        
1              147
2               95
3              125
4              103
5              149
6               74

 Now I'll build a dataset classifying them instead of showing these numbers.<br>
 I'll use the values shown by the 25% and 75% rows in the description table above to this.<br>
 Given a stat, if it's value equals or it's below the 25% value, it will be classified as **weak**.<br>
 On the other hand, if it's value equals or it's above the 75% value, it will be classified as **strong**.<br>
 Values that fall in the middle of these two marks will be classified as **standard**.<br>
 An overral ratting will be assigned to each Pokemon depending on it's stats. <br><br>
 **Weak** stats will count as -1, **Strong** stats will count as +1 and **Standard** stats will count as 0,<br>
 which will sum to an ratting between -6 and +6.

First i'll update the Stat Values Dataset with the information about Strong and Weak Stats

In [335]:
weak_hp = common_df["HP"].quantile(0.25)
strong_hp = common_df["HP"].quantile(0.75)

weak_attack = common_df["Attack"].quantile(0.25)
strong_attack = common_df["Attack"].quantile(0.75)

weak_defense = common_df["Defense"].quantile(0.25)
strong_defense = common_df["Defense"].quantile(0.75)

weak_spatk = common_df["Sp. Atk"].quantile(0.25)
strong_spatk = common_df["Sp. Atk"].quantile(0.75)

weak_spdef = common_df["Sp. Def"].quantile(0.25)
strong_spdef = common_df["Sp. Def"].quantile(0.75)

weak_speed = common_df["Speed"].quantile(0.25)
strong_speed = common_df["Speed"].quantile(0.75)

values = {"Value": ["weak","strong"], 
          "HP": [weak_hp, strong_hp], 
          "Attack": [weak_attack, strong_attack], 
          "Defense": [weak_defense, strong_defense], 
          "Sp. Atk": [weak_spatk, strong_spatk], 
          "Sp. Def": [weak_spdef, strong_spdef], 
          "Speed": [weak_speed, strong_speed]}

df = pd.DataFrame(values)

In [336]:
stat_values_df = stat_values_df.append(df)

In [337]:
stat_values_df.index = [0,1,2,3]

In [338]:
stat_values_df

Value     HP  Attack  Defense  Sp. Atk  Sp. Def  Speed
0     max  255.0   165.0    230.0    150.0    230.0  160.0
1     min    1.0     5.0      5.0     10.0     20.0    5.0
2    weak   50.0    52.0     50.0     45.0     50.0   45.0
3  strong   78.0    90.0     85.0     85.0     80.0   84.0

Now it's time to generate a new dataset with the Pokemons classified

In [349]:
common_classified_df = common_df.copy()

common_classified_df = common_classified_df.drop(columns=["Legendary"])

In [350]:
def classify_pokemon_dataframe(pk_df, stats_df):
    def get_stat_values_by_row_index(index):
        return {"HP": float(pk_df.loc[index, "HP"]), 
                "Attack": float(pk_df.loc[index, "Attack"]), 
                "Defense": float(pk_df.loc[index, "Defense"]), 
                "Sp. Atk": float(pk_df.loc[index, "Sp. Atk"]), 
                "Sp. Def": float(pk_df.loc[index, "Sp. Def"]), 
                "Speed": float(pk_df.loc[index, "Speed"])}
    
    def classify_stats_by_row_index(stats, index):
        for key in stats:
            weak, strong = get_weak_strong_values_by_stat(key)
            
            if stats[key] <= weak:
                pk_df.loc[index, key] = "weak"
            elif stats[key] > weak and stats[key] < strong:
                pk_df.loc[index, key] = "standard"
            elif stats[key] >= strong:
                pk_df.loc[index, key] = "strong"
    
    def rate_pokemon_by_row_index(stats, i):
        rating = 0
        weight = 1
        for key in stats:
            weak, strong = get_weak_strong_values_by_stat(key)
            if stats[key] <= weak:
                rating -= weight
            elif stats[key] >= strong:
                rating += weight
                
        return rating
    
    def get_weak_strong_values_by_stat(stat):
        return float(stats_df.loc[2, stat]), float(stats_df.loc[3, stat])
    
    ratings = []
    for i in pk_df.index:
        stats = get_stat_values_by_row_index(i)
        classify_stats_by_row_index(stats, i)
        rating = rate_pokemon_by_row_index(stats, i)
        ratings.append(rating)
    pk_df["Rating"] = ratings

In [351]:
classify_pokemon_dataframe(common_classified_df, stat_values_df)

In [352]:
common_classified_df

#                 Name  Type 1  Type 2        HP    Attack   Defense  \
0      1            Bulbasaur   Grass  Poison      weak      weak      weak   
1      2              Ivysaur   Grass  Poison  standard  standard  standard   
2      3             Venusaur   Grass  Poison    strong  standard  standard   
4      4           Charmander    Fire     NaN      weak      weak      weak   
5      5           Charmeleon    Fire     NaN  standard  standard  standard   
..   ...                  ...     ...     ...       ...       ...       ...   
787  711  GourgeistSuper Size   Ghost   Grass    strong    strong    strong   
788  712             Bergmite     Ice     NaN  standard  standard    strong   
789  713              Avalugg     Ice     NaN    strong    strong    strong   
790  714               Noibat  Flying  Dragon      weak      weak      weak   
791  715              Noivern  Flying  Dragon    strong  standard  standard   

      Sp. Atk   Sp. Def     Speed  Generation  Rating  
0    standard  standard      weak           1      -4  
1    standard    strong  standard           1       1  
2      strong    strong  standard           1       3  
4    standard      weak  standard           1      -4  
5    standard  standard  standard           1       0  
..        ...       ...       ...         ...     ...  
787  standard  standard  standard           6       3  
788      weak      weak      weak           6      -2  
789      weak      weak      weak           6       0  
790      weak      weak  standard           6      -5  
791    strong    strong    strong           6       4  

[693 rows x 12 columns]

Now I have a dataset good enough to start looking for the most powerful common pokeon based on stats only.<br>

### 3 - The Very Best

I'll begin my search with a simple question: What are the strongest pokemon in the dataset based on rating?<br>
More clearly: What are those with Rating of 5 or 6?

In [365]:
best_common_df = common_classified_df[common_classified_df["Rating"] >= 5]

In [366]:
best_common_df

#                  Name   Type 1    Type 2        HP    Attack  \
64    59              Arcanine     Fire       NaN    strong    strong   
161  149             Dragonite   Dragon    Flying    strong    strong   
165  151                   Mew  Psychic       NaN    strong    strong   
249  230               Kingdra    Water    Dragon  standard    strong   
267  248             Tyranitar     Rock      Dark    strong    strong   
271  251                Celebi  Psychic     Grass    strong    strong   
282  260              Swampert    Water    Ground    strong    strong   
313  289               Slaking   Normal       NaN    strong    strong   
408  373             Salamence   Dragon    Flying    strong    strong   
412  376             Metagross    Steel   Psychic    strong    strong   
493  445              Garchomp   Dragon    Ground    strong    strong   
548  490               Manaphy    Water       NaN    strong    strong   
696  635             Hydreigon     Dark    Dragon    strong    strong   
713  647  KeldeoOrdinary Forme    Water  Fighting    strong  standard   
714  647  KeldeoResolute Forme    Water  Fighting    strong  standard   
717  649              Genesect      Bug     Steel  standard    strong   

      Defense   Sp. Atk   Sp. Def     Speed  Generation  Rating  
64   standard    strong    strong    strong           1       5  
161    strong    strong    strong  standard           1       5  
165    strong    strong    strong    strong           1       6  
249    strong    strong    strong    strong           2       5  
267    strong    strong    strong  standard           2       5  
271    strong    strong    strong    strong           2       6  
282    strong    strong    strong  standard           3       5  
313    strong    strong  standard    strong           3       5  
408  standard    strong    strong    strong           3       5  
412    strong    strong    strong  standard           3       5  
493    strong  standard    strong    strong           4       5  
548    strong    strong    strong    strong           4       6  
696    strong    strong    strong    strong           5       6  
713    strong    strong    strong    strong           5       5  
714    strong    strong    strong    strong           5       5  
717    strong    strong    strong    strong           5       5

In [367]:
print("Very Best Total:",len(best_common_df.index))

Very Best Total: 16


Fortunately, sixteen pokemon are more than enough to build a team.<br>
There is also a small issue: I have consulted prestigious scholars on the subject which certified me that still some of these<br>
are Legendary. Seems like my original dataset have some errors.
I will generate a new dataset filtering out *Mew, Celebi, Manaphy, Kaldeo (both rows)* and *Genesect*

In [368]:
best_common_clean_df = best_common_df.drop(index=[165, 271, 548, 713, 714, 717])

In [369]:
best_common_clean_df

#       Name  Type 1   Type 2        HP  Attack   Defense   Sp. Atk  \
64    59   Arcanine    Fire      NaN    strong  strong  standard    strong   
161  149  Dragonite  Dragon   Flying    strong  strong    strong    strong   
249  230    Kingdra   Water   Dragon  standard  strong    strong    strong   
267  248  Tyranitar    Rock     Dark    strong  strong    strong    strong   
282  260   Swampert   Water   Ground    strong  strong    strong    strong   
313  289    Slaking  Normal      NaN    strong  strong    strong    strong   
408  373  Salamence  Dragon   Flying    strong  strong  standard    strong   
412  376  Metagross   Steel  Psychic    strong  strong    strong    strong   
493  445   Garchomp  Dragon   Ground    strong  strong    strong  standard   
696  635  Hydreigon    Dark   Dragon    strong  strong    strong    strong   

      Sp. Def     Speed  Generation  Rating  
64     strong    strong           1       5  
161    strong  standard           1       5  
249    strong    strong           2       5  
267    strong  standard           2       5  
282    strong  standard           3       5  
313  standard    strong           3       5  
408    strong    strong           3       5  
412    strong  standard           3       5  
493    strong    strong           4       5  
696    strong    strong           5       6

In [371]:
print("Very Best (Clean) Total:",len(best_common_clean_df.index))

Very Best (Clean) Total: 10


Sorting by types

In [378]:
best_common_clean_df.sort_values(by=["Type 1", "Type 2"])

#       Name  Type 1   Type 2        HP  Attack   Defense   Sp. Atk  \
696  635  Hydreigon    Dark   Dragon    strong  strong    strong    strong   
161  149  Dragonite  Dragon   Flying    strong  strong    strong    strong   
408  373  Salamence  Dragon   Flying    strong  strong  standard    strong   
493  445   Garchomp  Dragon   Ground    strong  strong    strong  standard   
64    59   Arcanine    Fire      NaN    strong  strong  standard    strong   
313  289    Slaking  Normal      NaN    strong  strong    strong    strong   
267  248  Tyranitar    Rock     Dark    strong  strong    strong    strong   
412  376  Metagross   Steel  Psychic    strong  strong    strong    strong   
249  230    Kingdra   Water   Dragon  standard  strong    strong    strong   
282  260   Swampert   Water   Ground    strong  strong    strong    strong   

      Sp. Def     Speed  Generation  Rating  
696    strong    strong           5       6  
161    strong  standard           1       5  
408    strong    strong           3       5  
493    strong    strong           4       5  
64     strong    strong           1       5  
313  standard    strong           3       5  
267    strong  standard           2       5  
412    strong  standard           3       5  
249    strong    strong           2       5  
282    strong  standard           3       5

What about the Types? A team of champions must be diverse

In [385]:
best_common_clean_df.groupby(["Type 1", "Type 2"]).count()["#"]

Type 1  Type 2 
Dark    Dragon     1
Dragon  Flying     2
        Ground     1
Rock    Dark       1
Steel   Psychic    1
Water   Dragon     1
        Ground     1
Name: #, dtype: int64

There are surely a good amount of dragons here.<br>

**Hydreigon** is the only Rating 6 Pokemon in the final list, it is a good idea to recruit it.<br>
If I try to pick any other non-dragon, what are my choices?

In [388]:
best_common_clean_df[(best_common_clean_df["Type 1"] != "Dragon") & (best_common_clean_df["Type 2"] != "Dragon")]

#       Name  Type 1   Type 2      HP  Attack   Defense Sp. Atk  \
64    59   Arcanine    Fire      NaN  strong  strong  standard  strong   
267  248  Tyranitar    Rock     Dark  strong  strong    strong  strong   
282  260   Swampert   Water   Ground  strong  strong    strong  strong   
313  289    Slaking  Normal      NaN  strong  strong    strong  strong   
412  376  Metagross   Steel  Psychic  strong  strong    strong  strong   

      Sp. Def     Speed  Generation  Rating  
64     strong    strong           1       5  
267    strong  standard           2       5  
282    strong  standard           3       5  
313  standard    strong           3       5  
412    strong  standard           3       5

Exactly Five!<br>
By adding all of these, the team is complete!<br>

In [392]:
champions_df = best_common_clean_df[best_common_clean_df["Name"].isin(["Hydreigon", "Arcanine", "Tyranitar", "Swampert", "Slaking", "Metagross"])]

In [393]:
champions_df

#       Name  Type 1   Type 2      HP  Attack   Defense Sp. Atk  \
64    59   Arcanine    Fire      NaN  strong  strong  standard  strong   
267  248  Tyranitar    Rock     Dark  strong  strong    strong  strong   
282  260   Swampert   Water   Ground  strong  strong    strong  strong   
313  289    Slaking  Normal      NaN  strong  strong    strong  strong   
412  376  Metagross   Steel  Psychic  strong  strong    strong  strong   
696  635  Hydreigon    Dark   Dragon  strong  strong    strong  strong   

      Sp. Def     Speed  Generation  Rating  
64     strong    strong           1       5  
267    strong  standard           2       5  
282    strong  standard           3       5  
313  standard    strong           3       5  
412    strong  standard           3       5  
696    strong    strong           5       6

### 4 - Conclusion

After fishing out of more than 800 pokemon for the most powerful common, based on status only,<br>
it would be nice to have this champion team on a file ready to be consulted.<br>
For this final list, the classification of stats and Rating used is not exactly useful, so i'd<br>
rather get the information on the original dataset, with the numbers

In [399]:
champions_final_df = pokemon_df[pokemon_df["Name"].isin(["Hydreigon", "Arcanine", "Tyranitar", "Swampert", "Slaking", "Metagross"])]
champions_final_df = champions_final_df.drop(columns=["Legendary"])

In [400]:
champions_final_df

#       Name  Type 1   Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  \
64    59   Arcanine    Fire      NaN   90     110       80      100       80   
267  248  Tyranitar    Rock     Dark  100     134      110       95      100   
282  260   Swampert   Water   Ground  100     110       90       85       90   
313  289    Slaking  Normal      NaN  150     160      100       95       65   
412  376  Metagross   Steel  Psychic   80     135      130       95       90   
696  635  Hydreigon    Dark   Dragon   92     105       90      125       90   

     Speed  Generation  
64      95           1  
267     61           2  
282     60           3  
313    100           3  
412     70           3  
696     98           5

In [405]:
name = "pokemon_champion_team.csv"
print("Saving file:", name)
champions_final_df.to_csv(name)

Saving file: pokemon_champion_team.csv


After a great result, we all know which ones to catch on the next gameplay